File: apps/worker/src/middleware/jwt.ts
import type { MiddlewareHandler } from "hono"

import type { AppEnv } from "../runtime/types"

// Minimal JWT implementation for Edge (no external deps)
interface JWTPayload {
  iat: number
  exp: number
  iss: string
}

async function createHmacKey(secret: string): Promise<CryptoKey> {
  const enc = new TextEncoder()
  return crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign", "verify"],
  )
}

function base64UrlEncode(data: ArrayBuffer | Uint8Array): string {
  const bytes = data instanceof Uint8Array ? data : new Uint8Array(data)
  let binary = ""
  for (const b of bytes) binary += String.fromCharCode(b)
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "")
}

function base64UrlDecode(str: string): Uint8Array {
  str = str.replace(/-/g, "+").replace(/_/g, "/")
  while (str.length % 4) str += "="
  const binary = atob(str)
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i)
  return bytes
}

export async function signJWT(payload: JWTPayload, secret: string): Promise<string> {
  const enc = new TextEncoder()
  const header = base64UrlEncode(enc.encode(JSON.stringify({ alg: "HS256", typ: "JWT" })))
  const body = base64UrlEncode(enc.encode(JSON.stringify(payload)))
  const data = `${header}.${body}`
  const key = await createHmacKey(secret)
  const sig = await crypto.subtle.sign("HMAC", key, enc.encode(data))
  return `${data}.${base64UrlEncode(sig)}`
}

export async function verifyJWT(token: string, secret: string): Promise<JWTPayload | null> {
  const parts = token.split(".")
  if (parts.length !== 3) return null

  const [header, body, sig] = parts
  const key = await createHmacKey(secret)
  const enc = new TextEncoder()
  const data = `${header}.${body}`
  const valid = await crypto.subtle.verify(
    "HMAC",
    key,
    base64UrlDecode(sig).buffer as ArrayBuffer,
    enc.encode(data),
  )
  if (!valid) return null

  const payload: JWTPayload = JSON.parse(new TextDecoder().decode(base64UrlDecode(body)))
  if (payload.exp && payload.exp < Math.floor(Date.now() / 1000)) return null

  return payload
}

export function generateToken(expiresMinutes: number) {
  const now = Math.floor(Date.now() / 1000)
  let exp: number
  if (expiresMinutes === 0) {
    exp = now + 15 * 60 // 15 minutes
  } else if (expiresMinutes === -1) {
    exp = now + 30 * 24 * 60 * 60 // 30 days
  } else {
    exp = now + expiresMinutes * 60
  }

  const payload: JWTPayload = { iat: now, exp, iss: "wheel" }
  const expireAt = new Date(exp * 1000).toISOString()
  return { payload, expireAt }
}

export const jwtAuth = (): MiddlewareHandler<AppEnv> => {
  return async (c, next) => {
    const authHeader = c.req.header("Authorization")
    if (!authHeader?.startsWith("Bearer ")) {
      return c.json({ success: false, error: "Unauthorized" }, 401)
    }

    const token = authHeader.slice(7)
    const secret = c.env.JWT_SECRET
    const payload = await verifyJWT(token, secret)
    if (!payload) {
      return c.json({ success: false, error: "Invalid or expired token" }, 401)
    }

    await next()
  }
}

File: apps/worker/internal/middleware/jwt.go
package middleware

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

// JWTPayload matches the TS JWT payload structure exactly.
type JWTPayload struct {
	IAT int64  `json:"iat"`
	EXP int64  `json:"exp"`
	ISS string `json:"iss"`
}

func base64UrlEncode(data []byte) string {
	s := base64.RawURLEncoding.EncodeToString(data)
	return s
}

func base64UrlDecode(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(s)
}

// SignJWT produces an HS256 JWT compatible with the TypeScript implementation.
func SignJWT(payload JWTPayload, secret string) (string, error) {
	header := base64UrlEncode([]byte(`{"alg":"HS256","typ":"JWT"}`))
	body, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	encodedBody := base64UrlEncode(body)
	data := header + "." + encodedBody

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(data))
	sig := base64UrlEncode(mac.Sum(nil))

	return data + "." + sig, nil
}

// VerifyJWT verifies an HS256 JWT and returns the payload.
func VerifyJWT(token, secret string) (*JWTPayload, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid token format")
	}

	data := parts[0] + "." + parts[1]
	sig, err := base64UrlDecode(parts[2])
	if err != nil {
		return nil, fmt.Errorf("invalid signature encoding")
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(data))
	expected := mac.Sum(nil)

	if !hmac.Equal(sig, expected) {
		return nil, fmt.Errorf("invalid signature")
	}

	body, err := base64UrlDecode(parts[1])
	if err != nil {
		return nil, fmt.Errorf("invalid payload encoding")
	}

	var payload JWTPayload
	if err := json.Unmarshal(body, &payload); err != nil {
		return nil, fmt.Errorf("invalid payload JSON")
	}

	if payload.EXP > 0 && payload.EXP < time.Now().Unix() {
		return nil, fmt.Errorf("token expired")
	}

	return &payload, nil
}

// GenerateToken creates a JWT payload and expireAt string.
// expiresMinutes: 0 = 15 minutes, -1 = 30 days, other = custom minutes.
func GenerateToken(expiresMinutes int) (JWTPayload, string) {
	now := time.Now().Unix()
	var exp int64
	switch {
	case expiresMinutes == 0:
		exp = now + 15*60
	case expiresMinutes == -1:
		exp = now + 30*24*60*60
	default:
		exp = now + int64(expiresMinutes)*60
	}

	payload := JWTPayload{IAT: now, EXP: exp, ISS: "wheel"}
	expireAt := time.Unix(exp, 0).UTC().Format(time.RFC3339)
	return payload, expireAt
}

// JWTAuth is a Gin middleware that validates Bearer JWT tokens.
func JWTAuth(jwtSecret string) gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization")
		if !strings.HasPrefix(authHeader, "Bearer ") {
			c.JSON(http.StatusUnauthorized, gin.H{"success": false, "error": "Unauthorized"})
			c.Abort()
			return
		}

		token := authHeader[7:]
		_, err := VerifyJWT(token, jwtSecret)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"success": false, "error": "Invalid or expired token"})
			c.Abort()
			return
		}

		c.Next()
	}
}
